### PR TITLE
feat: add CORS to SecurityPolicy

### PR DESCRIPTION
**Pull request by zhaohuabing**
What this PR does:
* Adds CORS to SecurityPolicy.
* Follow Golang name convention to rename  IR type "Cors" to "CORS"

Relate issues: 
* https://github.com/envoyproxy/gateway/issues/1845
* https://github.com/envoyproxy/gateway/issues/567

Related PRs:
* https://github.com/envoyproxy/gateway/pull/1957
* https://github.com/envoyproxy/gateway/pull/2016

BTW, the string match types are a bit messy. They're defined in multiple APIs but look similar. For most of EG APIs(Ratelimit and CORS for now, more in the future), the match types are just used to represent the [StringMatcher](https://www.envoyproxy.io/docs/envoy/latest/api-v3/type/matcher/v3/string.proto) XDS API. I think we can have one common string match structure used by all the EG APIs. I would like to raise a PR to refactor it if it makes sense.

### CODE REVIEW

File: api/v1alpha1/securitypolicy_types.go [hunk 1]
```diff
@@ -41,8 +41,74 @@ type SecurityPolicySpec struct {
...
+	//
+	// +kubebuilder:validation:MinLength=1
+	// +kubebuilder:validation:MaxLength=1024
+	Value string `json:"value"`
+
+	// IgnoreCase specifies whether the match should be case insensitive.
+	// This has no effect for the safe_regex match.
+	// Defaults to false.
+	// +optional
+	IgnoreCase bool `json:"caseSensitive,omitempty"`
```
Comment (arkodg) [inline api/v1alpha1/securitypolicy_types.go:83]: vote for rm `ignoreCase` now, and introduce later if needed

File: api/v1alpha1/securitypolicy_types.go [hunk 2]
```diff
@@ -41,8 +41,74 @@ type SecurityPolicySpec struct {
...
 	TargetRef gwapiv1a2.PolicyTargetReferenceWithSectionName `json:"targetRef"`
+
+	// CORS defines the configuration for Cross-Origin Resource Sharing (CORS).
+	CORS *CORS `json:"cors,omitempty"`
+}
+
+// CORS defines the configuration for Cross-Origin Resource Sharing (CORS).
+type CORS struct {
+	// AllowOrigins defines the origins that are allowed to make requests.
+	AllowOrigins []StringMatch `json:"allowOrigins,omitempty" yaml:"allowOrigins,omitempty"`
```
Comment (arkodg) [inline api/v1alpha1/securitypolicy_types.go:52]: min length for `AllowOrigins` should be 1

File: api/v1alpha1/securitypolicy_types.go [hunk 3]
```diff
@@ -41,8 +41,74 @@ type SecurityPolicySpec struct {
...
+	// MatchExact :the input string must match exactly the match value.
+	MatchExact MatchType = "Exact"
+
+	// MatchPrefix :the input string must start with the match value.
+	MatchPrefix MatchType = "Prefix"
+
+	// MatchSuffix :the input string must end with the match value.
+	MatchSuffix MatchType = "Suffix"
+
+	// MatchContains :the input string must contain the match value.
```
Comment (arkodg) [inline api/v1alpha1/securitypolicy_types.go:102]: vote to  rm `Contains`, dont think we should expose this yet

**Comment by arkodg** [comment]
looking good @zhaohuabing ! added some comments

File: internal/ir/xds.go [hunk 4]
```diff
@@ -675,11 +675,16 @@ type StringMatch struct {
...
+	// Contains match condition.
+	Contains *string `json:"contains,omitempty" yaml:"contains,omitempty"`
 	// SafeRegex match condition.
 	SafeRegex *string `json:"safeRegex,omitempty" yaml:"safeRegex,omitempty"`
 	// Distinct match condition.
 	// Used to match any and all possible unique values encountered within the Name field.
 	Distinct bool `json:"distinct" yaml:"distinct"`
+	// IgnoreCase specifies whether the match should be case insensitive.
+	// This has no effect for the safe_regex match.
+	IgnoreCase bool `json:"ignoreCase" yaml:"ignoreCase"`
```
Comment (arkodg) [inline internal/ir/xds.go:687]: can we also rm `IgnoreCase` from here

File: api/v1alpha1/securitypolicy_types.go [hunk 5]
```diff
@@ -41,8 +41,74 @@ type SecurityPolicySpec struct {
...
+	// CORS defines the configuration for Cross-Origin Resource Sharing (CORS).
+	CORS *CORS `json:"cors,omitempty"`
+}
+
+// CORS defines the configuration for Cross-Origin Resource Sharing (CORS).
+type CORS struct {
+	// AllowOrigins defines the origins that are allowed to make requests.
+	AllowOrigins []StringMatch `json:"allowOrigins,omitempty" yaml:"allowOrigins,omitempty"`
+	// AllowMethods defines the methods that are allowed to make requests.
+	AllowMethods []string `json:"allowMethods,omitempty" yaml:"allowMethods,omitempty"`
```
[the remark this answers is not in the document]
Comment (zhaohuabing) [reply api/v1alpha1/securitypolicy_types.go:54]: None, just like empty AllowOrigins.

**Review by arkodg — APPROVED** [review]
LGTM thanks !

**Comment by zhaohuabing** [comment]
Need to retrigger the coverage-test.

File: api/v1alpha1/securitypolicy_types.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (zhaohuabing) [reply api/v1alpha1/securitypolicy_types.go:56]: I did some  experiments, and it seems that CORS filter don't respect `AllowMethods` at all.

```
              routes:
              - match:
                  prefix: "/cors/open"
                route:
                  cluster: backend_service
                typed_per_filter_config:
                  envoy.filters.http.cors:
                    "@type": type.googleapis.com/envoy.extensions.filters.http.cors.v3.CorsPolicy
                    allow_methods: "POST"
```
curl -H "Origin: http://example.com" \
  -H "Access-Control-Request-Method: GET" \
  -H "Access-Control-Request-Headers: X-Requested-With" \
  -X OPTIONS --verbose \
  http://localhost:8002/cors/open
*   Trying 127.0.0.1:8002...
* Connected to localhost (127.0.0.1) port 8002 (#0)
> OPTIONS /cors/open HTTP/1.1
> Host: localhost:8002
> User-Agent: curl/8.1.2
> Accept: */*
> Origin: http://example.com
> Access-Control-Request-Method: GET
> Access-Control-Request-Headers: X-Requested-With
> 
< HTTP/1.1 200 OK
< access-control-allow-origin: http://example.com
< access-control-allow-methods: POST
< date: Wed, 25 Oct 2023 17:27:31 GMT
< server: envoy
< content-length: 0
< 
* Connection #0 to host localhost left intact
```

[the remark this answers is not in the document]
Comment (zhaohuabing) [reply api/v1alpha1/securitypolicy_types.go:56]: I did some experiments, the CORS filter won't set "Access-Control-Request-Headers" in the preflight response if AllowMethods is empty. So it means all Method Types are not allowed.

```
curl -H "Origin: http://example.com" \
  -H "Access-Control-Request-Method: GET" \
  -H "Access-Control-Request-Headers: X-Requested-With" \
  -X OPTIONS --verbose \
  http://localhost:8002/cors/open
*   Trying 127.0.0.1:8002...
* Connected to localhost (127.0.0.1) port 8002 (#0)
> OPTIONS /cors/open HTTP/1.1
> Host: localhost:8002
> User-Agent: curl/8.1.2
> Accept: */*
> Origin: http://example.com
> Access-Control-Request-Method: GET
> Access-Control-Request-Headers: X-Requested-With
> 
< HTTP/1.1 200 OK
< access-control-allow-origin: http://example.com
< date: Wed, 25 Oct 2023 17:32:39 GMT
< server: envoy
< content-length: 0
< 
* Connection #0 to host localhost left intact
```

[the remark this answers is not in the document]
Comment (arkodg) [reply api/v1alpha1/securitypolicy_types.go:56]: cool, so its opt in, thanks for running the test

File: api/v1alpha1/envoyproxy_metric_types.go [hunk 6]
```diff
@@ -60,7 +60,7 @@ type ProxyPrometheusProvider struct {
 }
 
 // Match defines the stats match configuration.
-type Match struct {
+type Match struct { // TODO: zhaohuabing this type should be renamed to StatsMatch
```
Comment (zirain) [inline api/v1alpha1/envoyproxy_metric_types.go:63]: can you open an issue to track this?

**Review by zirain — APPROVED** [review]

File: api/v1alpha1/envoyproxy_metric_types.go (hunk 6 above)
Comment (zhaohuabing) [reply api/v1alpha1/envoyproxy_metric_types.go:63]: https://github.com/envoyproxy/gateway/issues/2067

**Pull request merged**